### PR TITLE
feat: Support unnamed OP chains

### PR DESCRIPTION
**Pull request by BrianBland**
* Update `ChainSpec.is_optimism` to return true if the chain has the Bedrock hardfork configured (past or present), which allows arbitrary genesis files to be used for unnamed OP chains.
* Parse EIP-1559 elasticity parameters from Optimism genesis files instead of defaulting to using Ethereum's base fee params

### CODE REVIEW

File: crates/primitives/src/chain/spec.rs [hunk 1]
```diff
@@ -574,6 +574,11 @@ impl ChainSpec {
     /// Returns `true` if this chain contains Optimism configuration.
     #[inline]
     pub fn is_optimism(&self) -> bool {
+        #[cfg(feature = "optimism")]
+        if self.hardforks.contains_key(&Hardfork::Bedrock) {
+            return true;
+        };
```
Comment (mattsse) [inline crates/primitives/src/chain/spec.rs:580]: this seems reasonable, doesn't impact regular eth nodes

File: crates/primitives/src/chain/spec.rs [hunk 2]
```diff
@@ -1612,11 +1619,29 @@ struct OptimismGenesisInfo {
...
+                .as_object()
+                .and_then(|obj| obj.get("eip1559Denominator").and_then(|value| value.as_u64()))
+        });
+        let base_fee_params = if let (Some(elasticity), Some(denominator)) =
+            (eip1559_elasticity, eip1559_denominator)
+        {
+            BaseFeeParams::new(denominator as u128, elasticity as u128)
+        } else {
+            BaseFeeParams::ethereum()
+        };
```
Comment (mattsse) [inline crates/primitives/src/chain/spec.rs:1644]: this does the "optimism" lookup twice, I think we can nest this with an if let Some

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/primitives/src/chain/spec.rs (hunk 1 above)
Comment (BrianBland) [reply crates/primitives/src/chain/spec.rs:580]: I considered simply returning `false` if the `optimism` feature is disabled. Is it worth also making this change or simply keeping the previous name check?

File: crates/primitives/src/chain/spec.rs [hunk 3]
```diff
@@ -1612,11 +1619,29 @@ struct OptimismGenesisInfo {
...
     fn extract_from(genesis: &Genesis) -> Self {
+        let eip1559_elasticity = genesis.config.extra_fields.get("optimism").and_then(|value| {
+            value
+                .as_object()
+                .and_then(|obj| obj.get("eip1559Elasticity").and_then(|value| value.as_u64()))
+        });
+        let eip1559_denominator = genesis.config.extra_fields.get("optimism").and_then(|value| {
+            value
+                .as_object()
+                .and_then(|obj| obj.get("eip1559Denominator").and_then(|value| value.as_u64()))
```
Comment (DaniPopes) [inline crates/primitives/src/chain/spec.rs:1636]: can we un-nest by having
```
get
and_then as_object
get
and_then
```

File: crates/primitives/src/chain/spec.rs [hunk 4]
```diff
@@ -573,6 +573,14 @@ impl ChainSpec {
 
     /// Returns `true` if this chain contains Optimism configuration.
     #[inline]
+    #[cfg(feature = "optimism")]
```
Comment (BrianBland) [inline crates/primitives/src/chain/spec.rs:576]: Note: This change attempts to preserve the `const` behavior when possible, and also prioritizes the `const` `self.chain.is_optimism()` case over the slower runtime-evaluated hardfork check. The performance penalty should only impact unnamed chains or non-OP chains when the optimism feature is enabled.

File: crates/primitives/src/chain/spec.rs (hunk 1 above)
Comment (emhane) [reply crates/primitives/src/chain/spec.rs]: we are trying to get rid of the feature all together, so this is good the way it is https://github.com/paradigmxyz/reth/issues/7649

File: crates/net/discv5/src/network_stack_id.rs [hunk 5]
```diff
@@ -21,7 +21,7 @@ impl NetworkStackId {
     pub const OPSTACK: &'static [u8] = b"opstack";
 
     /// Returns the [`NetworkStackId`] that matches the given [`ChainSpec`].
-    pub const fn id(chain: &ChainSpec) -> Option<&'static [u8]> {
```
Comment (emhane) [inline crates/net/discv5/src/network_stack_id.rs:24]: pls revert

File: crates/primitives/src/chain/spec.rs (hunk 4 above)
Comment (emhane) [reply crates/primitives/src/chain/spec.rs:576]: nice

**Review by emhane — APPROVED** [review]

File: crates/net/discv5/src/network_stack_id.rs (hunk 5 above)
Comment (BrianBland) [reply crates/net/discv5/src/network_stack_id.rs:24]: This currently can only be `const` when the optimism feature is disabled. Do we want this to be defined conditionally?

Comment (emhane) [reply crates/net/discv5/src/network_stack_id.rs:24]: ok, thanks for the explanation

**Review by mattsse — APPROVED** [review]

**Pull request merged**